packages/asset-swapper/src/network/source_sampler.ts

```diff
@@ -0,0 +1,216 @@
+            return [];
+        }
+        const calls = await this._getSellQuoteCallsAsync(tokenAddressPath, takerFillAmounts);
+        return await Promise.all(calls.map(async c =>
+            c.getDexSamplesFromResult(
+                await this._sellContractHelper.ethCallAsync(
+                    this._sellContractFunction,
+                    c.args,
+                ),
+            ).filter(s => s.output),
+        ));
+    }
+
+    public async getBuySamplesAsync(
+        tokenAddressPath: string[],
+        makerFillAmounts: BigNumber[],
+    ): Promise<DexSample[][]> {
+        if (!this.canConvertTokens(tokenAddressPath)) {
+            return [];
+        }
+        const calls = await this._getBuyQuoteCallsAsync(tokenAddressPath, makerFillAmounts);
+        return await Promise.all(calls.map(async c =>
+            c.getDexSamplesFromResult(
+                await this._buyContractHelper.ethCallAsync(
+                    this._buyContractFunction,
+                    c.args,
+                ),
+            ).filter(s => s.output),
+        ));
+    }
+
+    public async getMultiHopSellCallInfosAsync(
+        tokenAddressPath: Address[],
+        takerFillAmount: BigNumber,
+        callOpts: Partial<ChainEthCallOpts> = {},
+    ): Promise<MultiHopCallInfo[]> {
+        if (!this.canConvertTokens(tokenAddressPath)) {
+            return [];
+        }
+        const calls = (await this._getSellQuoteCallsAsync(tokenAddressPath, [takerFillAmount]));
+        return calls.flat(1).map(c =>
+            createMultiHopCallInfo(
+                this._sellContractHelper,
+                this._sellContractFunction,
+                c.args,
+                (...args) => c.getDexSamplesFromResult(...args)[0],
+                callOpts,
+            ),
+        );
+    }
+
+    public async getMultiHopBuyCallInfosAsync(
+        tokenAddressPath: Address[],
+        makerFillAmount: BigNumber,
+        callOpts: Partial<ChainEthCallOpts> = {},
+    ): Promise<MultiHopCallInfo[]> {
+        if (!this.canConvertTokens(tokenAddressPath)) {
+            return [];
+        }
+        const calls = (await this._getBuyQuoteCallsAsync(tokenAddressPath, [makerFillAmount]));
+        return calls.flat(1).map(c =>
+            createMultiHopCallInfo(
+                this._buyContractHelper,
+                this._buyContractFunction,
+                c.args,
+                (...args) => c.getDexSamplesFromResult(...args)[0],
+                callOpts,
+            ),
+        );
+    }
+
+    protected abstract _getSellQuoteCallsAsync(
+        tokenAddressPath: string[],
+        takerFillAmounts: BigNumber[],
+    ): Promise<SamplerEthCall<TFillData, TSellSamplerFunctionArgs, TSellSamplerFunctionReturn>[]>;
+
+    protected abstract _getBuyQuoteCallsAsync(
+        tokenAddressPath: string[],
+        takerFillAmounts: BigNumber[],
+    ): Promise<SamplerEthCall<TFillData, TBuySamplerFunctionArgs, TBuySamplerFunctionReturn>[]>;
```

Comment:
Most source samplers just need to override these two functions, and maybe `canConvertTokens()`.